FREEWARE/v10/imagemagick/display.man
.ad l
.nh
.TH display 1 "1 May 1994" "ImageMagick"
.SH NAME
display - display an image on any workstation running X
.SH SYNOPSIS
.B "display" [ \fIoptions\fP ...] \fIfile\fP
[ [ \fIoptions\fP ...] \fIfile\fP ...]
.SH DESCRIPTION
\fBDisplay\fP is a machine architecture independent image processing
and display program.  It can display an image on any workstation
display running an X server.  \fBDisplay\fP can read and write many of
the more popular image formats (e.g. JPEG, TIFF, PNM, etc.).  You can
perform these functions on the image:

    o display information about the image
    o write the image to a file
    o print the image to a Postscript printer
    o delete the image file
    o load an image from a file
    o display the next image
    o display the former image
    o display a sequence of images as a slide show
    o select the image to display by its thumbnail rather than name
    o undo last image transformation
    o restore the image to its original size
    o refresh the image
    o toggle the colormap type: : Shared or Private
    o half the image size
    o double the image size
    o resize the image
    o trim the image edges
    o crop the image
    o cut the image
    o flop image in the horizontal direction
    o flip image in the vertical direction
    o rotate the image 90 degrees clockwise
    o rotate the image 90 degrees counter-clockwise
    o rotate the image
    o shear the image
    o invert the colors of the image
    o vary the color brightness
    o vary the color saturation
    o vary the image hue
    o gamma correct the image
    o sharpen the image contrast
    o dull the image contrast
    o perform histogram equalization on the image
    o perform histogram normalization on the image
    o negate the image colors
    o reduce the speckles within an image
    o eliminate peak noise from an image
    o detect edges within the image
    o emboss an image
    o oil paint an image
    o convert the image to grayscale
    o set the maximum number of unique colors in the image
    o annotate the image with text
    o add a border to the image
    o composite image with another
    o edit an image pixel color
    o edit the image matte information
    o add an image comment
    o display information about this program
    o discard all images and exit program
    o change the level of magnification
.SH EXAMPLES
To scale an image of a cockatoo to exactly 640 pixels in width and 480
pixels in height and position the window at location (200,200), use:
.PP
.B
     display -geometry 640x480\+200\+200! cockatoo.miff
.PP
To display an image of a cockatoo without a border centered on a
backdrop, use:
.PP
.B
     display +borderwidth -backdrop cockatoo.miff
.PP
To tile an image of a cockatoo onto the root window, use:
.PP
.B
     display -window root cockatoo.miff
.PP
To display a visual image directory of all your JPEG images, use:
.PP
.B
     display 'vid:*.jpg'
.PP
To display a MAP image that is 640 pixels in width and 480
pixels in height with 256 colors, use:
.PP
.B
     display -size 640x480\+256 cockatoo.map
.SH OPTIONS
.TP 5
.B "-backdrop"
display the image centered on a backdrop.

This backdrop covers the entire workstation screen and is useful for
hiding other X window activity while viewing the image.   The color of
the backdrop is specified as the background color.  Refer to \fBX
RESOURCES\fP for details.
.TP 5
.B "-blur"
blur an image.
.TP 5
.B "-border \fI<width>x<height>fP"
surround the image with a border or color.  See \fBX(1)\fP for details
about the geometry specification.
 
The color of the border is obtained from the X server and is defined as
\fBbordercolor\fP (class \fBborderColor\fP).  See \fBX(1)\fP for details.
.TP 5
.B "-colormap \fItype\fP"
the type of colormap: \fBShared\fP or \fBPrivate\fP.

This option only applies when the default X server visual is
\fIPseudoColor\fP or \fIGrayScale\fP.  Refer to \fB-visual\fP for more
details.  By default, a shared colormap is allocated.  The image shares
colors with other X clients.  Some image colors could be approximated,
therefore your image may look very different than intended.  Choose
\fBPrivate\fP and the image colors appear exactly as they are
defined.  However, other clients may go "technicolor" when the image
colormap is installed.
.TP 5
.B "-colors \fIvalue\fP"
preferred number of colors in the image.

The actual number of colors in the image may be less than your request,
but never more.  Note, this is a color reduction option.  Images with
less unique colors than specified with this option will remain unchanged.
Refer to \fBquantize(9)\fP for more details.

Note, options \fB-dither\fP, \fB-colorspace\fP, and \fB-treedepth\fP affect
the color reduction algorithm.
.TP 5
.B "-colorspace \fIvalue\fP"
the type of colorspace: \fBGRAY\fP, \fBOHTA\fP, \fBRGB\fP, \fBXYZ\fP,
\fBYCbCr\fP, \fBYIQ\fP, \fBYPbPr\fP, or \fBYUV\fP.

Color reduction, by default, takes place in the RGB color space.
Empirical evidence suggests that distances in color spaces such as YUV
or YIQ correspond to perceptual color differences more closely
than do distances in RGB space.  These color spaces may give better
results when color reducing an image.  Refer to \fBquantize(9)\fP for
more details.

The \fB-colors\fP or \fB-monochrome\fP option is required for this option
to take effect.
.TP 5
.B "-comment \fIstring\fP"
annotate an image with a comment.

By default, each image is commented with its file name.  Use this
option to assign a specific comment to the image.  Optionally you can
include the image filename, type, width, height, or scene number by
embedding special format characters.  Embed \fB%f\fP for filename,
\fB%m\fP for magick, \fB%w\fP for width, \fB%h\fP for height, \fB%s\fP
for scene number, or \fB\\n\fP for newline.  For example,

.nf
     -comment "%m:%f %wx%h"
.fi

produces an image comment of \fBMIFF:bird.miff 512x480\fP for an image
titled \fBbird.miff\fP and whose width is 512 and height is 480.
 
If the first character of \fIstring\fP is \fB@\fP, the image comment is read
from a file titled by the remaining characters in the string.
.TP 5
.B "-compress \fItype\fP"
the type of image compression: \fIQEncoded\fP or \fIRunlengthEncoded\fP.

Use this option with \fB-write\fP to specify the the type of image
compression.  See \fBmiff(5)\fP for details.

Specify \fB\+compress\fP to store the binary image in an uncompressed format.
The default is the compression type of the specified image file.
.TP 5
.B "-contrast"
enhance or reduce the image contrast.

This option enhances the intensity differences between the
lighter and darker elements of the image.  Use \fB-contrast\fP to
enhance the image or \fB+contrast\fP to reduce the image contrast.
.TP 5
.B "-crop \fI<width>x<height>{\+-}<x offset>{\+-}<y offset>\fP"
preferred size and location of the cropped image.  See \fBX(1)\fP for details
about the geometry specification.

Use cropping to apply image processing options to, or display, a
particular area of an image.  Use \fB-crop 0x0\fP to remove edges that are
the background color.

The equivalent X resource for this option is \fBcropGeometry\fP
(class \fBCropGeometry\fP).  See \fBX RESOURCES\fP for details.
.TP 5
.B "-delay \fIseconds\fP"
display the next image after pausing.

This option is useful when viewing several images in sequence.  Each
image will display and wait the number of seconds specified before the
next image is displayed.  The default is to display the image
and wait until you choose to display the next image or terminate the
program.
.TP 5
.B "-density \fI<width>x<height>
vertical and horizontal density of the image.

This option specifies an image density for a Postscript page.
The default is 72 dots per inch in the horizontal and vertical direction.
Use this option to alter the default density.
.TP 5
.B "-despeckle"
reduce the speckles within an image.
.TP 5
.B "-display \fIhost:display[.screen]\fP"
specifies the X server to contact; see \fBX(1)\fP.
.TP 5
.B "-dither"
apply Floyd/Steinberg error diffusion to the image.

The basic strategy of dithering is to trade intensity resolution for
spatial resolution by averaging the intensities of several neighboring
pixels.  Images which suffer from severe contouring when reducing colors
can be improved with this option.

The \fB-colors\fP or \fB-monochrome\fP option is required
for this option to take effect.
.TP 5
.B "-edge"
detect edges within an image.
.TP 5
.B "-enhance"
apply a digital filter to enhance a noisy image.
.TP 5
.B "-equalize"
perform histogram equalization to the image.
.TP 5
.B "-flip"
create a "mirror image" by reflecting the image scanlines in the vertical
direction.
.TP 5
.B "-flop"
create a "mirror image" by reflecting the image scanlines in the horizontal
direction.
.TP 5
.B "-gamma \fIvalue\fP"
level of gamma correction.

The same color image displayed on two different workstations may look
different due to differences in the display monitor.  Use gamma
correction to adjust for this color difference.  Reasonable values
extend from 0.8 to 2.3.

You can apply separate gamma values to the red, green, and blue
channels of the image with a gamma value list delineated with commas
(i.e. 1.7,2.3,1.2).
.TP 5
.B "-geometry \fI<width>{%}x<height>{%}{\+-}<x offset>{\+-}<y offset>{!}\fP"
preferred size and location of the image window.  See \fBX(1)\fP for details
about the geometry specification.  By default, the window size is the image
size and the location is choosen by you when it is mapped.

By default, the width and height are maximum values.  That is, the
image is expanded or contracted to fit the width and height value while
maintaining the aspect ratio of the image.  Append an exclamation point
to the geometry to force the image size to exactly the size you
specify.  For example, if you specify \fB640x480!\fP the image width is
set to 640 pixels and height to 480.  If only one factor is
specified, both the width and height assume the value. 

To specify a percentage width or height instead, append \fB%\fP.  The
image size is multiplied by the width and height percentages to obtain
the final image dimensions.  To increase the size of an image, use a
value greater than 100 (e.g. 125%).  To decrease an image's size, use a
percentage less than 100.

When displaying an image on an X server, \fI<x offset>\fP and
\fI<y offset>\fP is relative to the root window.

The equivalent X resource for this option is \fBimageGeometry\fP
(class \fBImageGeometry\fP).  See \fBX RESOURCES\fP for details.
.TP 5
.B "-interlace \fItype\fP"
the type of interlacing scheme: \fBNONE\fP, \fBLINE\fP, or \fBPLANE\fP.

This option is used to specify the type of interlacing scheme for raw
image formats such as \fBRGB\fP or \fBYUV\fP.  \fBNONE\fP means do not
interlace (RGBRGBRGBRGBRGBRGB...), \fBLINE\fP uses scanline
interlacing (RRR...GGG...BBB...RRR...GGG...BBB...), and \fBPLANE\fP uses
plane interlacing (RRRRRR...GGGGGG...BBBBBB...).
.TP 5
.B "-label \fIname\fP"
assign a label to an image.

Use this option to assign a specific label to the image.  Optionally
you can include the image filename, type, width, height, or scene
number in the label by embedding special format characters.   Embed
\fB%f\fP for filename, \fB%m\fP for magick, \fB%w\fP for width,
\fB%h\fP for height, or \fB%s\fP for scene number.  For example,
.nf
     -label "%m:%f %wx%h"
.fi
produces an image label of \fBMIFF:bird.miff 512x480\fP for an image
titled \fBbird.miff\fP and whose width is 512 and height is 480.

If the first character of \fIstring\fP is \fB@\fP, the image label is read
from a file titled by the remaining characters in the string.
 
When converting to Postscript, use this option to specify a header string
to print above the image.
.TP 5
.B "-map \fItype\fP"
display image using this Standard Colormap type.

Choose from these Standard Colormap types:

    best
    default
    gray
    red
    green
    blue

The X server must support the Standard Colormap you choose, otherwise an
error occurs.  Use \fBlist\fP as the type and \fBdisplay(1)\fP searches

UNIX/SOURCE/Xw/Xw/Display.c
/*************************************<+>*************************************
 *****************************************************************************
 **
 **   File:        Display.c
 **
 **   Project:     X Widgets
 **
 **   Description: Code for TextEdit widget ascii sink
 **
 *****************************************************************************
 **   
 **   Copyright (c) 1988 by Hewlett-Packard Company
 **   Copyright (c) 1987, 1988 by Digital Equipment Corporation, Maynard,
 **             Massachusetts, and the Massachusetts Institute of Technology,
 **             Cambridge, Massachusetts
 **   
 **   Permission to use, copy, modify, and distribute this software 
 **   and its documentation for any purpose and without fee is hereby 
 **   granted, provided that the above copyright notice appear in all 
 **   copies and that both that copyright notice and this permission 
 **   notice appear in supporting documentation, and that the names of 
 **   Hewlett-Packard, Digital or  M.I.T.  not be used in advertising or 
 **   publicity pertaining to distribution of the software without 
 **   written prior permission.
 **   
 **   DIGITAL DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
 **   ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
 **   DIGITAL BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
 **   ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 **   WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
 **   ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
 **   SOFTWARE.
 **   
 *****************************************************************************
 *************************************<+>*************************************/


#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>
#include <X11/Intrinsic.h>
#include <X11/IntrinsicP.h>
#include <X11/StringDefs.h>
#include <Xw/Xw.h>
#include <Xw/XwP.h>

#include <Xw/TextEditP.h>

#ifdef DEBUG
#include <stdio.h>
#endif DEBUG


#define GETLASTPOS (*(source->scan))(source, 0, XwstLast, XwsdRight, 1, TRUE)
/* Private Ascii TextSink Definitions */

static unsigned bufferSize = 200;

typedef struct _AsciiSinkData {
    Pixel foreground;
    GC normgc, invgc, xorgc;
    XFontStruct *font;
    int tabwidth;
    Pixmap insertCursorOn;
    XwInsertState laststate;
} AsciiSinkData, *AsciiSinkPtr;

static unsigned char *buf;

/* XXX foreground default should be XtDefaultFGPixel. How do i do that?? */

static XtResource SinkResources[] = {
    {XtNfont, XtCFont, XtRFontStruct, sizeof (XFontStruct *),
        XtOffset(AsciiSinkPtr, font), XtRString, "Fixed"},
    {XtNforeground, XtCForeground, XtRPixel, sizeof (int),
        XtOffset(AsciiSinkPtr, foreground), XtRString, "XtDefaultForeground"},    
};

/* Utilities */

/*--------------------------------------------------------------------------+*/
static int CharWidth (data, x, margin, c)
/*--------------------------------------------------------------------------+*/
  AsciiSinkData *data;
  int x;
  int margin ;
  unsigned char c;
{
    int     width, nonPrinting;
    XFontStruct *font = data->font;

    if (c == '\t')
        /* This is totally bogus!! need to know tab settings etc.. */
	return data->tabwidth - ((x-margin) % data->tabwidth);
    if (c == LF)
	c = SP;
    nonPrinting = (c < SP);
    if (nonPrinting) c += '@';

    if (font->per_char &&
	    (c >= font->min_char_or_byte2 && c <= font->max_char_or_byte2))
	width = font->per_char[c - font->min_char_or_byte2].width;
    else
	width = font->min_bounds.width;

    if (nonPrinting)
	width += CharWidth(data, x, margin, '^');

    return width;
}

/* Sink Object Functions */

#define LBEARING(x) \
    ((font->per_char != NULL && \
      ((x) >= font->min_char_or_byte2 && (x) <= font->max_char_or_byte2)) \
	? font->per_char[(x) - font->min_char_or_byte2].lbearing \
	: font->min_bounds.lbearing)

/*--------------------------------------------------------------------------+*/
static int AsciiDisplayText (w, x, y, pos1, pos2, highlight)
/*--------------------------------------------------------------------------+*/
  Widget w;
  Position x, y;
  int highlight;
  XwTextPosition pos1, pos2;
{
    XwTextSink *sink = ((XwTextEditWidget)w)->text.sink;
    XwTextSource *source = ((XwTextEditWidget)w)->text.source;
    AsciiSinkData *data = (AsciiSinkData *) sink->data ;
    int margin = ((XwTextEditWidget)w)->text.leftmargin ;

    XFontStruct *font = data->font;
    int     j, k;
    Dimension width;
    XwTextBlock blk;
    GC gc = highlight ? data->invgc : data->normgc;
    GC invgc = highlight ? data->normgc : data->invgc;

    y += font->ascent;
    j = 0;
    while (pos1 < pos2) {
	pos1 = (*(source->read))(source, pos1, &blk, pos2 - pos1);
	for (k = 0; k < blk.length; k++) {
	    if (j >= bufferSize - 5) {
		bufferSize *= 2;
		buf = (unsigned char *) XtRealloc(buf, bufferSize);
	    }
	    buf[j] = blk.ptr[k];
	    if (buf[j] == LF)
		buf[j] = ' ';
	    else if (buf[j] == '\t') {
	        XDrawImageString(XtDisplay(w), XtWindow(w),
			gc, x - LBEARING(*buf), y, buf, j);
		buf[j] = 0;
		x += XTextWidth(data->font, buf, j);
		width = CharWidth(data, x, margin, '\t');
		XFillRectangle(XtDisplay(w), XtWindow(w), invgc, x,
			       y - font->ascent, width,
			       (Dimension) (data->font->ascent +
					    data->font->descent));
		x += width;
		j = -1;
	    }
	    else
		if (buf[j] < ' ') {
		    buf[j + 1] = buf[j] + '@';
		    buf[j] = '^';
		    j++;
		}
	    j++;
	}
    }
    XDrawImageString(XtDisplay(w), XtWindow(w), gc, x - LBEARING(*buf), y,
		     buf, j);
}


#   define insertCursor_width 6
#   define insertCursor_height 3
static char insertCursor_bits[] = {0x0c, 0x1e, 0x33};

/*--------------------------------------------------------------------------+*/
static Pixmap CreateInsertCursor(s)
/*--------------------------------------------------------------------------+*/
Screen *s;
{

    return (XCreateBitmapFromData (DisplayOfScreen(s), RootWindowOfScreen(s),
        insertCursor_bits, insertCursor_width, insertCursor_height));
}

/*
 * The following procedure manages the "insert" cursor.
 */

/*--------------------------------------------------------------------------+*/
static AsciiInsertCursor (w, x, y, state)
/*--------------------------------------------------------------------------+*/
  Widget w;
  Position x, y;
  XwInsertState state;
{
    XwTextSink *sink = ((XwTextEditWidget)w)->text.sink;
    AsciiSinkData *data = (AsciiSinkData *) sink->data;

/*
    XCopyArea(sink->dpy,
	      (state == XwisOn) ? data->insertCursorOn : data->insertCursorOff, w,
	      data->normgc, 0, 0, insertCursor_width, insertCursor_height,
	      x - (insertCursor_width >> 1), y - (insertCursor_height));

    if (state != data->laststate && XtIsRealized(w))
	XCopyPlane(XtDisplay(w),
		  data->insertCursorOn, XtWindow(w),
		  data->xorgc, 0, 0, insertCursor_width, insertCursor_height,
		  x - (insertCursor_width >> 1), y - (insertCursor_height), 1);
*/
/* This change goes with the cursor hack for the broken server */
    XCopyArea (XtDisplay(w), data->insertCursorOn, XtWindow (w),
	       data->xorgc, 0, 0, insertCursor_width, insertCursor_height,
	       x - (insertCursor_width >> 1), y - (insertCursor_height));

    data->laststate = state;
}

/*
 * Clear the passed region to the background color.
 */

/*--------------------------------------------------------------------------+*/
static AsciiClearToBackground (w, x, y, width, height)
/*--------------------------------------------------------------------------+*/
  Widget w;
  Position x, y;
  Dimension width, height;
{
    XwTextSink *sink = ((XwTextEditWidget)w)->text.sink;
    AsciiSinkData *data = (AsciiSinkData *) sink->data;
    XFillRectangle(XtDisplay(w), XtWindow(w), data->invgc, x, y, width, height);
}

/*
 * Given two positions, find the distance between them.
 */

/*--------------------------------------------------------------------------+*/
static AsciiFindDistance (w, fromPos, fromx, toPos,
			  resWidth, resPos, resHeight)
/*--------------------------------------------------------------------------+*/
  Widget w;
  XwTextPosition fromPos;	/* First position. */
  int fromx;			/* Horizontal location of first position. */
  XwTextPosition toPos;		/* Second position. */
  int *resWidth;		/* Distance between fromPos and resPos. */
  int *resPos;			/* Actual second position used. */
  int *resHeight;		/* Height required. */
{
    XwTextSink *sink = ((XwTextEditWidget)w)->text.sink;
    XwTextSource *source = ((XwTextEditWidget)w)->text.source;
    int margin = ((XwTextEditWidget)w)->text.leftmargin ;

    AsciiSinkData *data;
    register    XwTextPosition index, lastPos;
    register unsigned char   c;
    XwTextBlock blk;

    data = (AsciiSinkData *) sink->data;
    /* we may not need this */
    lastPos = GETLASTPOS;
    (*(source->read))(source, fromPos, &blk, toPos - fromPos);
    *resWidth = 0;
    for (index = fromPos; index != toPos && index < lastPos; index++) {
	if (index - blk.firstPos >= blk.length)
	    (*(source->read))(source, index, &blk, toPos - fromPos);
	c = blk.ptr[index - blk.firstPos];
	if (c == LF) {
	    *resWidth += CharWidth(data, fromx + *resWidth, margin, SP);
	    index++;
	    break;
	}
	*resWidth += CharWidth(data, fromx + *resWidth, margin, c);
    }
    *resPos = index;
    *resHeight = data->font->ascent + data->font->descent;
}


/*--------------------------------------------------------------------------+*/
static TextFit AsciiTextFit (w, fromPos, fromx, width, wrap, wrapWhiteSpace,
			     fitPos, drawPos, nextPos, resWidth, resHeight)
/*--------------------------------------------------------------------------+*/
  Widget w;
  XwTextPosition fromPos; 	/* Starting position. */
  int fromx;			/* Horizontal location of starting position. */
  int width;			/* Desired width. */
  int wrap;			/* Whether line should wrap at all */
  int wrapWhiteSpace;		/* Whether line should wrap at white space */

  XwTextPosition *fitPos ;	/* pos of last char which fits in specified
				   width */
  XwTextPosition *drawPos ;	/* pos of last char to draw in specified
				   width based on wrap model */
  XwTextPosition *nextPos ;	/* pos of next char to draw outside specified
				   width based on wrap model */
  int *resWidth;		/* Actual width used. */
  int *resHeight;		/* Height required. */
{
    XwTextSink *sink = ((XwTextEditWidget)w)->text.sink;
    XwTextSource *source = ((XwTextEditWidget)w)->text.source;
    int margin = ((XwTextEditWidget)w)->text.leftmargin ;
    AsciiSinkData *data;
    XwTextPosition lastPos, pos, whiteSpacePosition;
    XwTextPosition fitL, drawL ;
    		/* local equivalents of fitPos, drawPos, nextPos */
    int     lastWidth, whiteSpaceWidth, whiteSpaceSeen ;
    int useAll ;
    TextFit fit ;
    unsigned char    c;
    XwTextBlock blk;
    
    data = (AsciiSinkData *) sink->data;
    lastPos = GETLASTPOS;

    *resWidth = 0;
    *fitPos = fromPos ;
    *drawPos = -1 ;
    c = 0;
    useAll = whiteSpaceSeen = FALSE ;

    pos = fromPos ;
    fitL = pos - 1 ;
    (*(source->read))(source, fromPos, &blk, bufferSize);

    while (*resWidth <= width)
      {	lastWidth = *resWidth;
	fitL = pos - 1 ;
	if (pos >= lastPos)
	  { pos = lastPos ;
	    fit = tfEndText ;
	    useAll = TRUE ;
	    break ;
	  } ;
	if (pos - blk.firstPos >= blk.length)
	  (*(source->read))(source, pos, &blk, bufferSize);
	c = blk.ptr[pos - blk.firstPos];

	if (isNewline(c))
	  { fit = tfNewline ;
	    useAll = TRUE ;
	    break ;
	  }

	if (wrapWhiteSpace && isWhiteSpace(c))
	  { whiteSpaceSeen = TRUE ;
	    drawL = pos - 1 ;
	    whiteSpaceWidth = *resWidth;
	  } ;
	  
	*resWidth += CharWidth(data, fromx + *resWidth, margin, c);
	pos++ ;
    } /* end while */

    *fitPos = fitL ;
    *drawPos = fitL ;
    if (useAll)
      {
	*nextPos = pos + 1 ;
	*resWidth = lastWidth ;
      }
    else if (wrapWhiteSpace && whiteSpaceSeen)
      { *drawPos = drawL ;
	*nextPos = drawL + 2 ;
	*resWidth = whiteSpaceWidth ;
	fit = tfWrapWhiteSpace ;
      }
    else if (wrap)
      {
	*nextPos = fitL + 1 ;
	*resWidth = lastWidth ;
	fit = tfWrapAny ;
      }
    else
      {
	/* scan source for newline or end */
	*nextPos =
	  (*(source->scan)) (source, pos, XwstEOL, XwsdRight, 1, TRUE) + 1 ;
	*resWidth = lastWidth ;
	fit = tfNoFit ;
      }
    *resHeight = data->font->ascent + data->font->descent;
    return (fit) ;
}

/*--------------------------------------------------------------------------+*/
static AsciiFindPosition(w, fromPos, fromx, width, stopAtWordBreak, 
			 resPos, resWidth, resHeight)
/*--------------------------------------------------------------------------+*/
  Widget w;
  XwTextPosition fromPos; 	/* Starting position. */
  int fromx;			/* Horizontal location of starting position. */
  int width;			/* Desired width. */
  int stopAtWordBreak;		/* Whether the resulting position should be at
				   a word break. */
  XwTextPosition *resPos;	/* Resulting position. */
  int *resWidth;		/* Actual width used. */
  int *resHeight;		/* Height required. */
{
    XwTextSink *sink = ((XwTextEditWidget)w)->text.sink;
    XwTextSource *source = ((XwTextEditWidget)w)->text.source;
    int margin = ((XwTextEditWidget)w)->text.leftmargin ;
    AsciiSinkData *data;
    XwTextPosition lastPos, index, whiteSpacePosition;
    int     lastWidth, whiteSpaceWidth;
    Boolean whiteSpaceSeen;
    unsigned char c;
    XwTextBlock blk;
    data = (AsciiSinkData *) sink->data;
    lastPos = GETLASTPOS;

    (*(source->read))(source, fromPos, &blk, bufferSize);
    *resWidth = 0;
    whiteSpaceSeen = FALSE;
    c = 0;
    for (index = fromPos; *resWidth <= width && index < lastPos; index++) {
	lastWidth = *resWidth;
	if (index - blk.firstPos >= blk.length)
	    (*(source->read))(source, index, &blk, bufferSize);
	c = blk.ptr[index - blk.firstPos];
	if (c == LF) {
	    *resWidth += CharWidth(data, fromx + *resWidth, margin, SP);
	    index++;
	    break;
	}
	*resWidth += CharWidth(data, fromx + *resWidth, margin, c);
	if ((c == SP || c == TAB) && *resWidth <= width) {
	    whiteSpaceSeen = TRUE;
	    whiteSpacePosition = index;
	    whiteSpaceWidth = *resWidth;
	}
    }
    if (*resWidth > width && index > fromPos) {
	*resWidth = lastWidth;
	index--;
	if (stopAtWordBreak && whiteSpaceSeen) {
	    index = whiteSpacePosition + 1;
	    *resWidth = whiteSpaceWidth;
	}
    }
    if (index == lastPos && c != LF) index = lastPos + 1;
    *resPos = index;
    *resHeight = data->font->ascent + data->font->descent;
}


/*--------------------------------------------------------------------------+*/
static int AsciiResolveToPosition (w, pos, fromx, width,
				   leftPos, rightPos)
/*--------------------------------------------------------------------------+*/
  Widget w;
  XwTextPosition pos;
  int fromx,width;
  XwTextPosition *leftPos, *rightPos;
{
    int     resWidth, resHeight;
    XwTextSink *sink = ((XwTextEditWidget)w)->text.sink;
    XwTextSource *source = ((XwTextEditWidget)w)->text.source;

    AsciiFindPosition(w, pos, fromx, width, FALSE,
	    leftPos, &resWidth, &resHeight);
    if (*leftPos > GETLASTPOS)
	*leftPos = GETLASTPOS;
    *rightPos = *leftPos;
}


/*--------------------------------------------------------------------------+*/
static int AsciiMaxLinesForHeight (w)
/*--------------------------------------------------------------------------+*/
  Widget w;
{
    AsciiSinkData *data;
    XwTextSink *sink = ((XwTextEditWidget)w)->text.sink;

    data = (AsciiSinkData *) sink->data;
    return (int) ((w->core.height
		   - ((XwTextEditWidget) w)->text.topmargin
		   - ((XwTextEditWidget) w)->text.bottommargin
		   )
		  / (data->font->ascent + data->font->descent)
		  );
}


/*--------------------------------------------------------------------------+*/
static int AsciiMaxHeightForLines (w, lines)
/*--------------------------------------------------------------------------+*/
  Widget w;
  int lines;
{
    AsciiSinkData *data;
    XwTextSink *sink = ((XwTextEditWidget)w)->text.sink;

    data = (AsciiSinkData *) sink->data;
    return(lines * (data->font->ascent + data->font->descent));
}


/***** Public routines *****/

static Boolean initialized = FALSE;
static XContext asciiSinkContext;

/*--------------------------------------------------------------------------+*/
AsciiSinkInitialize()
/*--------------------------------------------------------------------------+*/
{
    if (initialized)
    	return;
    initialized = TRUE;

    asciiSinkContext = XUniqueContext();

    buf = (unsigned char *) XtMalloc(bufferSize);
}

/*--------------------------------------------------------------------------+*/
static Boolean XwAsciiSinkCheckData(self)
/*--------------------------------------------------------------------------+*/
  XwTextSink *self ;
{
  XwTextEditWidget tew = self->parent ;
  
  /* make sure margins are big enough to keep traversal highlight
     from obscuring text or cursor.
     */
  { int minBorder ;
    minBorder = (tew->primitive.traversal_type == XwHIGHLIGHT_OFF)
                  ? RequiredCursorMargin
		  : RequiredCursorMargin + tew->primitive.highlight_thickness ;

    if (tew->text.topmargin < minBorder)
      tew->text.topmargin = minBorder ;
    if (tew->text.bottommargin < minBorder)
      tew->text.bottommargin = minBorder ;
    if (tew->text.rightmargin < minBorder)
      tew->text.rightmargin = minBorder ;
    if (tew->text.leftmargin < minBorder)
      tew->text.leftmargin = minBorder ;
  }

  if ((*(self->maxLines))(tew) < 1)
    XtWarning("TextEdit window too small to display a single line of text.");

}

/*--------------------------------------------------------------------------+*/
void XwAsciiSinkDestroy (sink)
/*--------------------------------------------------------------------------+*/
    XwTextSink *sink;
{
    AsciiSinkData *data;
    data = (AsciiSinkData *) sink->data;
    XtFree((char *) data);
    XtFree((char *) sink);
}

/*--------------------------------------------------------------------------+*/
XwTextSink *XwAsciiSinkCreate (w, args, num_args)
/*--------------------------------------------------------------------------+*/
    Widget w;
    ArgList 	args;
    Cardinal 	num_args;
{
    XwTextSink *sink;
    AsciiSinkData *data;
    unsigned long valuemask = (GCFont | GCGraphicsExposures |
			       GCForeground | GCBackground | GCFunction);
    XGCValues values;
    unsigned long wid;
    XFontStruct *font;

    if (!initialized)
    	AsciiSinkInitialize();

    sink                    = XtNew(XwTextSink);
    sink->parent            = (XwTextEditWidget) w ; 
    sink->parent->text.sink = sink ;		/* disgusting */
    sink->display           = AsciiDisplayText;
    sink->insertCursor      = AsciiInsertCursor;
    sink->clearToBackground = AsciiClearToBackground;
    sink->findPosition      = AsciiFindPosition;
    sink->textFitFn         = AsciiTextFit;
    sink->findDistance      = AsciiFindDistance;
    sink->resolve           = AsciiResolveToPosition;
    sink->maxLines          = AsciiMaxLinesForHeight;
    sink->maxHeight         = AsciiMaxHeightForLines;
    sink->resources         = SinkResources;
    sink->resource_num      = XtNumber(SinkResources);
    sink->check_data        = XwAsciiSinkCheckData;
    sink->destroy           = XwAsciiSinkDestroy;
    sink->LineLastWidth	    = 0 ;
    sink->LineLastPosition  = 0 ;
    data                    = XtNew(AsciiSinkData);
    sink->data              = (int *)data;

    XtGetSubresources (w, (caddr_t)data, "display", "Display", 
		       SinkResources, XtNumber(SinkResources),
		       args, num_args);

/* XXX do i have to XLoadQueryFont or does the resource guy do it for me */

    font = data->font;
    values.function = GXcopy;
    values.font = font->fid;
    values.graphics_exposures = (Bool) FALSE;
    values.foreground = data->foreground;
    values.background = w->core.background_pixel;
    data->normgc = XtGetGC(w, valuemask, &values);
    values.foreground = w->core.background_pixel;
    values.background = data->foreground;
    data->invgc = XtGetGC(w, valuemask, &values);
    values.function = GXxor;
    values.foreground = data->foreground ^ w->core.background_pixel;
    values.background = 0;
    data->xorgc = XtGetGC(w, valuemask, &values);

    wid = -1;
    if ((!XGetFontProperty(font, XA_QUAD_WIDTH, &wid)) || wid <= 0) {
	if (font->per_char && font->min_char_or_byte2 <= '0' &&
	    		      font->max_char_or_byte2 >= '0')
	    wid = font->per_char['0' - font->min_char_or_byte2].width;
	else
	    wid = font->max_bounds.width;
    }
    if (wid <= 0) wid = 1;
    data->tabwidth = 8 * wid;
    data->font = font;

/*    data->insertCursorOn = CreateInsertCursor(XtScreen(w)); */
    
    {/* Correction from AsciiSink.c on R2 tape */
      Screen *screen = XtScreen(w);
      Display *dpy = XtDisplay(w);
      Window root = RootWindowOfScreen(screen);
      Pixmap bitmap = XCreateBitmapFromData(dpy, root, insertCursor_bits,
					    insertCursor_width,
					    insertCursor_height);
      Pixmap pixmap = XCreatePixmap(dpy, root,insertCursor_width,
				    insertCursor_height,
				    DefaultDepthOfScreen(screen));
      XGCValues gcv;
      GC gc;
      
      gcv.function = GXcopy;
      gcv.foreground = data->foreground ^ w->core.background_pixel;
      gcv.background = 0;
      gcv.graphics_exposures = False;
      gc = XtGetGC(w, (GCFunction | GCForeground | GCBackground |
		       GCGraphicsExposures), &gcv);
      XCopyPlane(dpy, bitmap, pixmap, gc, 0, 0, insertCursor_width,
		 insertCursor_height, 0, 0, 1);
      XtDestroyGC(gc);
      data->insertCursorOn = pixmap;
    }

    data->laststate = XwisOff;
    (*(sink->check_data))(sink);

    return sink;
}

